pool-manager.js

```diff
@@ -75,76 +84,120 @@ module.exports = {
-        // A ring buffer, holding the timestamp of each error. As we loop around the ring, the timestamp in the slot we're
-        // about to fill will tell us the error rate. That is, `maxFailure` number of requests in how many milliseconds?
-        endpoint.buffer = new Array(maxFailures - 1);
-        endpoint.bufferPointer = 0;
-      },
-      onEndpointReturned: function (endpoint, err) {
-        if (err) {
-          var oldestErrorTime, now;
-          if (endpoint.state === OPEN) {
-            return;
-          }
-
-          if (endpoint.buffer.length === 0) {
-            disableEndpoint(endpoint);
-            return;
+    function getRollingWindowConfiguration() {
+      return {
+        isInPool: isInPool,
+        onEndpointSelected: onEndpointSelected,
+        onEndpointRegistered: function (endpoint) {
+          endpoint.state = CLOSED;
+          // A ring buffer, holding the timestamp of each error. As we loop around the ring, the timestamp in the slot we're
+          // about to fill will tell us the error rate. That is, `maxFailure` number of requests in how many milliseconds?
+          endpoint.buffer = new RingBuffer(maxFailures - 1);
+        },
+        onEndpointReturned: function (endpoint, err) {
+          if (err) {
+            if (endpoint.state === OPEN) {
+              return;
+            }
+
+            if (endpoint.buffer.size === 0) {
+              disableEndpoint(endpoint);
+              return;
+            }
+
+            var now = Date.now();
+            var oldestErrorTime = endpoint.buffer.read();
+            endpoint.buffer.write(now);
+
+            if (endpoint.state === HALF_OPEN_PENDING || (oldestErrorTime != null && now - oldestErrorTime <= failureWindow)) {
+              disableEndpoint(endpoint);
+            }
+          } else if (endpoint.state === HALF_OPEN_PENDING) {
+            endpoint.state = CLOSED;
           }
+        }
+      };
+    }
 
-          oldestErrorTime = endpoint.buffer[endpoint.bufferPointer];
-          now = Date.now();
-          endpoint.buffer[endpoint.bufferPointer] = now;
-          endpoint.bufferPointer++;
-          endpoint.bufferPointer %= endpoint.buffer.length;
-
-          if (endpoint.state === HALF_OPEN_PENDING || (oldestErrorTime != null && now - oldestErrorTime <= failureWindow)) {
+    function getRateConfiguration() {
+      var maxErrorCount = failureRate * failureRateWindow;
+      return {
+        isInPool: isInPool,
+        onEndpointSelected: onEndpointSelected,
+        onEndpointRegistered: function (endpoint) {
+          endpoint.state = CLOSED;
+          endpoint.buffer = new RingBuffer(failureRateWindow);
+          endpoint.errors = 0;
+        },
+        onEndpointReturned: function (endpoint, err) {
+          var state = endpoint.state;
+          var newStatus = err ? 1 : 0;
+          var oldestStatus = endpoint.buffer.read() ? 1 : 0;
+          endpoint.buffer.write(!!newStatus);
```

Comment:
Casting to Boolean doesn't seem necessary, especially given that we cast Boolean back to 0/1 when reading the last status